Panic in taiko_payload_building

#### System information

Geth
Version: 1.11.5-stable
Git Commit: 9da8845f59f8a69e3f1df05903b1eaddc4ca4756
Git Commit Date: 20230605
Architecture: amd64
Go Version: go1.20.3
Operating System: linux
GOPATH=
GOROOT=

#### Expected behaviour

The execution engine doesn't panic.


#### Actual behaviour

The execution engine panics.


#### Steps to reproduce the behaviour


#### Backtrace

````
INFO [06-07|17:40:05.757] Chain head was updated                   number=14521 hash=192ba2..e1e887 root=3d9ba2..e89e2e elapsed="442.108µs"
INFO [06-07|17:40:06.174] Starting work on payload                 id=0xb003b09b18227f3c
INFO [06-07|17:40:06.174] Updated payload                          id=0xb003b09b18227f3c number=14522 hash=38ef18..ff594e txs=5    gas=1,588,501 fees=0 root=e7d790..d2d1bf
INFO [06-07|17:40:06.175] Stopping work on payload                 id=0xb003b09b18227f3c reason=delivery
panic: send on closed channel

goroutine 2012892 [running]:
github.com/ethereum/go-ethereum/miner.(*Payload).SetFullBlock.func1()
        github.com/ethereum/go-ethereum/miner/taiko_payload_building.go:21 +0x91
created by github.com/ethereum/go-ethereum/miner.(*Payload).SetFullBlock
        github.com/ethereum/go-ethereum/miner/taiko_payload_building.go:16 +0xfe
````

--- a/miner/taiko_payload_building.go
+++ b/miner/taiko_payload_building.go
@@ -13,17 +13,7 @@ func (payload *Payload) SetFullBlock(block *types.Block, fees *big.Int) {
 	payload.lock.Lock()
 	defer payload.lock.Unlock()
 
-	go func() {
-		payload.lock.Lock()
-		defer payload.lock.Unlock()
-
-		select {
-		case <-payload.done:
-			log.Info("SetFullBlock payload done received", "id", payload.id)
-		case payload.stop <- struct{}{}:
-		default:
-		}
-	}()
+	go payload.afterSetFullBlock()
 
 	payload.full = block
 	payload.fullFees = fees
@@ -34,4 +24,21 @@ func (payload *Payload) SetFullBlock(block *types.Block, fees *big.Int) {
 		"root", block.Root())
 
 	payload.cond.Broadcast() // fire signal for notifying full block
+}
+
+func (payload *Payload) afterSetFullBlock() {
+	payload.lock.Lock()
+	defer payload.lock.Unlock()
+
+	select {
+	case <-payload.done:
+		log.Info("SetFullBlock payload done received", "id", payload.id)
+		return
+	default:
+	}
+
+	select {
+	case payload.stop <- struct{}{}:
+	default:
+	}
 }

--- a/miner/taiko_payload_building_test.go
+++ b/miner/taiko_payload_building_test.go
@@ -53,6 +53,63 @@ func Test_SetFullBlock_AvoidPanic(t *testing.T) {
 	payload.SetFullBlock(block, fees)
 }
 
+func Test_AfterSetFullBlock_Panic_DoneChannelNotSent(t *testing.T) {
+	var (
+		db        = rawdb.NewMemoryDatabase()
+		recipient = common.HexToAddress("0xdeadbeef")
+	)
+	w, b := newTestWorker(t, params.TestChainConfig, ethash.NewFaker(), db, 0)
+	defer w.close()
+
+	timestamp := uint64(time.Now().Unix())
+	args := &BuildPayloadArgs{
+		Parent:       b.chain.CurrentBlock().Hash(),
+		Timestamp:    timestamp,
+		Random:       common.Hash{},
+		FeeRecipient: recipient,
+	}
+	payload, err := w.buildPayload(args)
+	if err != nil {
+		t.Fatalf("Failed to build payload %v", err)
+	}
+
+	// dont send on done channel, but close stop channel.
+	// should panic when sent on.
+	close(payload.stop)
+
+	assert.Panics(t, func() {
+		payload.afterSetFullBlock()
+	})
+}
+
+func Test_AfterSetFullBlock_AvoidPanic_DoneChannelSent(t *testing.T) {
+	var (
+		db        = rawdb.NewMemoryDatabase()
+		recipient = common.HexToAddress("0xdeadbeef")
+	)
+	w, b := newTestWorker(t, params.TestChainConfig, ethash.NewFaker(), db, 0)
+	defer w.close()
+
+	timestamp := uint64(time.Now().Unix())
+	args := &BuildPayloadArgs{
+		Parent:       b.chain.CurrentBlock().Hash(),
+		Timestamp:    timestamp,
+		Random:       common.Hash{},
+		FeeRecipient: recipient,
+	}
+	payload, err := w.buildPayload(args)
+	if err != nil {
+		t.Fatalf("Failed to build payload %v", err)
+	}
+
+	payload.done <- struct{}{}
+	close(payload.stop)
+
+	assert.NotPanics(t, func() {
+		payload.afterSetFullBlock()
+	})
+}
+
 func Test_SetFullBlock(t *testing.T) {
 	var (
 		db        = rawdb.NewMemoryDatabase()

